extensions/amp-story-360/0.1/amp-story-360.js

```diff
@@ -202,15 +276,204 @@ export class AmpStory360 extends AMP.BaseElement {
-          false /* callToInitialize */
+          StateProperty.GYROSCOPE_PERMISSION_STATE,
+          (permissionState) => this.onPermissionState_(permissionState)
+        );
+
+        storeService.subscribe(
+          StateProperty.CURRENT_PAGE_ID,
+          (currPageId) => {
+            this.isOnActivePage_ = currPageId === this.getPageId_();
+          },
+          true /* callToInitialize */
+        );
+      }),
+
+      Services.localizationServiceForOrNull(this.element).then(
+        (localizationService) => {
+          this.localizationService_ = localizationService;
+        }
+      ),
+    ]).then(() => {
+      attr('controls') === 'gyroscope' && this.checkGyroscopePermissions_();
+      return Promise.resolve();
+    });
+  }
+
+  /**
+   * @private
+   * @return {string} the page id
+   */
+  getPageId_() {
+    if (this.pageId_ == null) {
+      this.pageId_ = closest(dev().assertElement(this.element), (el) => {
+        return el.tagName.toLowerCase() === 'amp-story-page';
+      }).getAttribute('id');
+    }
+    return this.pageId_;
+  }
+
+  /**
+   * @param {string} permissionState
+   * @private
+   */
+  onPermissionState_(permissionState) {
+    if (permissionState === 'granted') {
+      this.enableGyroscope_();
+    } else if (permissionState === 'denied') {
+      this.gyroscopeControls_ = false;
+      this.togglePermissionClass_(true);
+    }
+  }
+
+  /** @private */
+  checkGyroscopePermissions_() {
+    //  If browser doesn't support DeviceOrientationEvent.
+    if (typeof DeviceOrientationEvent === 'undefined') return;
+
+    // If browser doesn't require permission for DeviceOrientationEvent.
+    if (typeof DeviceOrientationEvent.requestPermission === 'undefined')
+      this.enableGyroscope_();
+
+    // If permissions needed for DeviceOrientationEvent.
+    if (
+      typeof this.win.DeviceOrientationEvent.requestPermission === 'function'
+    ) {
+      this.win.DeviceOrientationEvent.requestPermission()
+        .catch(() => {
+          // If permissions weren't set, render activate button.
+          this.renderActivateButton_();
+        })
+        .then((permissionState) => {
+          // If permissions already set, set permission state.
+          permissionState && this.setPermissionState_(permissionState);
+        });
+    }
+  }
+
+  /**
+   * Creates a device orientation listener and sets gyroscopeControls_ state.
+   * If listener is not called in 1000ms, remove listener and resume animation.
+   * This happens on desktop browsers that support deviceorientation but aren't in motion.
+   * @private
+   */
+  enableGyroscope_() {
+    this.gyroscopeControls_ = true;
+    this.togglePermissionClass_(true);
+
+    const checkNoMotion = this.timer_.delay(() => {
+      this.gyroscopeControls_ = false;
+      if (this.isReady_ && this.isPlaying_) {
+        this.animate_();
+      }
+    }, 1000);
+
+    let rafTimeout;
+
+    this.win.addEventListener('deviceorientation', (e) => {
```

Comment:
Can you create a ticket to stop listening when the page becomes inactive? It should be easy to do but this PR is already a (good) big chunk of code :))

Reply:
For sure: #30000